File: backend/downloader.py
import subprocess
import threading
import re
import os
import uuid
import glob
import time
import imageio_ffmpeg

class DownloadTask:
    def __init__(self, task_id, url, quality="1080p"):
        self.task_id = task_id
        self.url = url
        self.quality = quality
        self.status = "pending"   # pending / downloading / processing / completed / error
        self.progress = 0.0
        self.filename = ""
        self.error = ""
        self.title = ""
        self.speed = ""
        self.eta = ""
        self.filesize = ""
        self.last_progress_time = time.time()

# --- 错误分类 ---
ERROR_MESSAGES = {
    "Video is private": "视频为私有，无法下载",
    "Sign in to confirm your age": "年龄限制视频，请在项目根目录放置 cookies.txt",
    "This video is not available": "视频不可用，可能已被删除",
    "Video unavailable": "视频不可用",
    "HTTP Error 403": "访问被拒绝，可能是地区限制或代理问题",
    "HTTP Error 404": "视频不存在或已被删除",
    "HTTP Error 429": "请求过于频繁，请稍后重试",
    "Network error": "网络连接失败，请检查代理设置",
    "Too Many Requests": "请求过于频繁，请稍后重试",
    "urlopen error": "网络连接失败，请检查网络或代理",
    "SSL": "SSL 证书错误，请检查网络环境",
    "timeout": "连接超时，请检查网络",
    "Proxy": "代理连接失败，请检查代理设置",
    "ERROR: Sign in": "需要登录，请提供 cookies.txt",
    "This live event will begin": "直播尚未开始",
    "This video is only available": "该视频有地区限制",
}

def classify_error(error_msg: str) -> str:
    """将 yt-dlp 错误信息映射为友好中文提示"""
    for keyword, friendly_msg in ERROR_MESSAGES.items():
        if keyword.lower() in error_msg.lower():
            return friendly_msg
    if len(error_msg) > 200:
        return f"下载失败：{error_msg[:200]}..."
    return f"下载失败：{error_msg}"

def sanitize_filename(name: str) -> str:
    """消毒文件名，移除危险字符"""
    # 移除路径遍历字符
    name = name.replace("..", "").replace("/", "").replace("\\", "")
    # 移除 Windows 非法字符
    name = re.sub(r'[<>:"|?*]', '', name)
    # 限制长度
    if len(name) > 200:
        name = name[:200]
    return name.strip() or "video"

# --- 画质格式映射 ---
QUALITY_FORMATS = {
    "1080p": "bv*[height<=1080][vcodec^=avc1]+ba[ext=m4a]/bv*[height<=1080]+ba[ext=m4a]/bv*+ba/b",
    "720p": "bv*[height<=720][vcodec^=avc1]+ba[ext=m4a]/bv*[height<=720]+ba[ext=m4a]/bv*+ba/b",
    "480p": "bv*[height<=480][vcodec^=avc1]+ba[ext=m4a]/bv*[height<=480]+ba[ext=m4a]/bv*+ba/b",
    "audio": "ba[ext=m4a]/ba/b",
}

DOWNLOAD_TIMEOUT = 300  # 5 分钟无进度视为超时

class VideoDownloader:
    def __init__(self, download_dir="downloads"):
        self.download_dir = os.path.abspath(download_dir)
        os.makedirs(self.download_dir, exist_ok=True)
        self.tasks: dict[str, DownloadTask] = {}
        self.ffmpeg_path = imageio_ffmpeg.get_ffmpeg_exe()
        self._cleanup_partial_files()
        self._lock = threading.Lock()

    def _cleanup_partial_files(self):
        """清理未合并的中间文件"""
        patterns = [
            os.path.join(self.download_dir, "*.f*.mp4"),
            os.path.join(self.download_dir, "*.f*.m4a"),
            os.path.join(self.download_dir, "*.f*.webm"),
            os.path.join(self.download_dir, "*.ytdl"),
            os.path.join(self.download_dir, "*.part"),
        ]
        for pattern in patterns:
            for f in glob.glob(pattern):
                try:
                    os.remove(f)
                except Exception:
                    pass

    def get_info(self, url: str, proxy: str = None) -> dict:
        """获取视频信息"""
        cmd = [
            "yt-dlp", "--dump-json", "--no-download", "--no-warnings",
            "--no-check-certificates", "--ffmpeg-location", self.ffmpeg_path,
        ]
        if proxy:
            cmd += ["--proxy", proxy]
        cmd.append(url)

        try:
            result = subprocess.run(cmd, capture_output=True, text=True, timeout=30, encoding="utf-8")
            if result.returncode == 0:
                import json
                info = json.loads(result.stdout)
                return {
                    "title": info.get("title", "未知"),
                    "duration": info.get("duration", 0),
                    "thumbnail": info.get("thumbnail", ""),
                    "uploader": info.get("uploader", "未知"),
                    "view_count": info.get("view_count", 0),
                }
            else:
                return {"error": classify_error(result.stderr)}
        except Exception as e:
            return {"error": classify_error(str(e))}

    def start_download(self, url: str, proxy: str = None, quality: str = "1080p") -> str:
        """启动下载任务，返回 task_id"""
        task_id = str(uuid.uuid4())[:8]
        task = DownloadTask(task_id, url, quality)
        with self._lock:
            self.tasks[task_id] = task

        thread = threading.Thread(target=self._download_worker, args=(task, proxy), daemon=True)
        thread.start()
        return task_id

    def get_status(self, task_id: str) -> dict:
        """获取下载状态"""
        task = self.tasks.get(task_id)
        if not task:
            return {"error": "任务不存在"}
        return {
            "task_id": task.task_id,
            "status": task.status,
            "progress": round(task.progress, 1),
            "filename": task.filename,
            "title": task.title,
            "speed": task.speed,
            "eta": task.eta,
            "error": task.error,
            "filesize": task.filesize,
            "quality": task.quality,
        }

    def list_files(self) -> list[dict]:
        """列出已下载的文件"""
        files = []
        try:
            for f in os.listdir(self.download_dir):
                filepath = os.path.join(self.download_dir, f)
                if os.path.isfile(filepath):
                    if re.search(r'\.f\d+\.\w+$', f) or f.endswith('.ytdl') or f.endswith('.part'):
                        continue
                    size_bytes = os.path.getsize(filepath)
                    if size_bytes < 1024:
                        size_str = f"{size_bytes} B"
                    elif size_bytes < 1024 * 1024:
                        size_str = f"{size_bytes / 1024:.1f} KB"
                    else:
                        size_str = f"{size_bytes / (1024*1024):.1f} MB"
                    files.append({"name": f, "size": size_str, "size_bytes": size_bytes})
            files.sort(key=lambda x: x["size_bytes"], reverse=True)
        except Exception:
            pass
        return files

    def active_task_count(self) -> int:
        """返回活跃任务数"""
        return sum(1 for t in self.tasks.values() if t.status in ("pending", "downloading", "processing"))

    def _download_worker(self, task: DownloadTask, proxy: str = None):
        """后台线程执行下载"""
        fmt = QUALITY_FORMATS.get(task.quality, QUALITY_FORMATS["1080p"])

        cmd = [
            "yt-dlp", "--newline", "--no-warnings",
            "--ffmpeg-location", self.ffmpeg_path,
            "--merge-output-format", "mp4",
            "-f", fmt,
            "--format-sort", "vcodec:h264,res,ext:mp4:m4a",
            "-o", os.path.join(self.download_dir, "%(title)s.%(ext)s"),
            "--user-agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36",
            "--no-part", "--retries", "30", "--fragment-retries", "30",
            "--retry-sleep", "linear=1::2", "--no-check-certificates",
            "--socket-timeout", "30", "--extractor-retries", "5",
            "--file-access-retries", "5", "--continue",
        ]

        cookies_file = os.path.join(os.path.dirname(self.download_dir), "cookies.txt")
        if os.path.exists(cookies_file):
            cmd += ["--cookies", cookies_file]

        if proxy:
            cmd += ["--proxy", proxy]
        cmd.append(task.url)

        task.status = "downloading"
        task.last_progress_time = time.time()

        # 超时检测线程
        def _timeout_checker():
            while task.status in ("downloading", "processing"):
                if time.time() - task.last_progress_time > DOWNLOAD_TIMEOUT:
                    task.status = "error"
                    task.error = "下载超时，可能是网络问题，请检查代理设置后重试"
                    return
                time.sleep(10)

        timeout_thread = threading.Thread(target=_timeout_checker, daemon=True)
        timeout_thread.start()

        try:
            process = subprocess.Popen(
                cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
                text=True, encoding="utf-8", errors="replace",
            )

            for line in process.stdout:
                line = line.strip()
                if line:
                    self._parse_progress(task, line)

            process.wait()

            if process.returncode == 0:
                task.status = "completed"
                task.progress = 100.0
                self._find_downloaded_file(task)
            else:
                task.status = "error"
                if not task.error:
                    task.error = classify_error(f"yt-dlp 返回错误码: {process.returncode}")

        except Exception as e:
            task.status = "error"
            task.error = classify_error(str(e))

    def _parse_progress(self, task: DownloadTask, line: str):
        """解析 yt-dlp 输出，提取进度"""
        # [download] 45.2% of 120.50MiB at 2.30MiB/s ETA 00:30
        match = re.search(r'\[download\]\s+([\d.]+)%\s+of\s+~?\s*(\S+)\s+at\s+(\S+)\s+ETA\s+(\S+)', line)
        if match:
            task.progress = float(match.group(1))
            task.filesize = match.group(2)
            task.speed = match.group(3)
            task.eta = match.group(4)
            task.status = "downloading"
            task.last_progress_time = time.time()
            return

        # [download] 100% of 120.50MiB
        match = re.search(r'\[download\]\s+([\d.]+)%\s+of\s+~?\s*(\S+)', line)
        if match:
            task.progress = float(match.group(1))
            task.filesize = match.group(2)
            task.last_progress_time = time.time()
            return

        # [download] Destination: filename.mp4
        match = re.search(r'\[download\] Destination:\s+(.+)', line)
        if match:
            task.filename = sanitize_filename(os.path.basename(match.group(1)))
            return

        # [Merger] Merging formats...
        if '[Merger]' in line or 'Merging' in line:
            task.status = "processing"
            task.progress = 99.0
            task.last_progress_time = time.time()
            return

        # already been downloaded
        match = re.search(r'\[download\]\s+(.+)\s+has already been downloaded', line)
        if match:
            task.filename = sanitize_filename(os.path.basename(match.group(1)))
            task.status = "completed"
            task.progress = 100.0
            return

        # ERROR
        if 'ERROR' in line:
            task.error = classify_error(line)

    def _find_downloaded_file(self, task: DownloadTask):
        """找到最新下载的文件"""
        if task.filename and not re.search(r'\.f\d+\.\w+$', task.filename):
            return
        try:
            files = [f for f in os.listdir(self.download_dir)
                     if f.endswith('.mp4') and not re.search(r'\.f\d+\.mp4$', f)]
            if files:
                files.sort(key=lambda f: os.path.getmtime(os.path.join(self.download_dir, f)), reverse=True)
                task.filename = files[0]
        except Exception:
            pass

File: src/downloader.js
const { spawn } = require('child_process');
const path = require('path');
const fs = require('fs');
const crypto = require('crypto');
const logger = require('./logger');

// 错误分类映射
const ERROR_MESSAGES = {
    'Video is private': '视频为私有，无法下载',
    'Sign in to confirm your age': '年龄限制视频，请在项目根目录放置 cookies.txt',
    'This video is not available': '视频不可用，可能已被删除',
    'Video unavailable': '视频不可用',
    'HTTP Error 403': '访问被拒绝，可能是地区限制或代理问题',
    'HTTP Error 404': '视频不存在或已被删除',
    'HTTP Error 429': '请求过于频繁，请稍后重试',
    'Network error': '网络连接失败，请检查代理设置',
    'Too Many Requests': '请求过于频繁，请稍后重试',
    'urlopen error': '网络连接失败，请检查网络或代理',
    'SSL': 'SSL 证书错误，请检查网络环境',
    'timeout': '连接超时，请检查网络',
    'Proxy': '代理连接失败，请检查代理设置',
    'ERROR: Sign in': '需要登录，请提供 cookies.txt',
    'This live event will begin': '直播尚未开始',
    'This video is only available': '该视频有地区限制',
};

// 画质格式映射
const QUALITY_FORMATS = {
    '1080p': 'bv*[height<=1080][vcodec^=avc1]+ba[ext=m4a]/bv*[height<=1080]+ba[ext=m4a]/bv*+ba/b',
    '720p': 'bv*[height<=720][vcodec^=avc1]+ba[ext=m4a]/bv*[height<=720]+ba[ext=m4a]/bv*+ba/b',
    '480p': 'bv*[height<=480][vcodec^=avc1]+ba[ext=m4a]/bv*[height<=480]+ba[ext=m4a]/bv*+ba/b',
    'audio': 'ba[ext=m4a]/ba/b',
};

const DOWNLOAD_TIMEOUT = 5 * 60 * 1000; // 5 分钟

function classifyError(msg) {
    const lower = msg.toLowerCase();
    for (const [keyword, friendly] of Object.entries(ERROR_MESSAGES)) {
        if (lower.includes(keyword.toLowerCase())) {
            return friendly;
        }
    }
    return msg.length > 200 ? `下载失败：${msg.substring(0, 200)}...` : `下载失败：${msg}`;
}

function sanitizeFilename(name) {
    return name
        .replace(/\.\./g, '')
        .replace(/[/\\]/g, '')
        .replace(/[<>:"|?*]/g, '')
        .substring(0, 200)
        .trim() || 'video';
}

class DownloadTask {
    constructor(taskId, url, quality = '1080p') {
        this.taskId = taskId;
        this.url = url;
        this.quality = quality;
        this.status = 'pending'; // pending / downloading / processing / completed / error
        this.progress = 0;
        this.filename = '';
        this.error = '';
        this.title = '';
        this.speed = '';
        this.eta = '';
        this.filesize = '';
        this.lastProgressTime = Date.now();
        this._timeoutTimer = null;
    }
}

class VideoDownloader {
    constructor(downloadDir = 'downloads') {
        this.downloadDir = path.resolve(downloadDir);
        if (!fs.existsSync(this.downloadDir)) {
            fs.mkdirSync(this.downloadDir, { recursive: true });
        }
        this.tasks = new Map();
        this.ytdlpPath = this._findYtdlp();
        this._cleanupPartialFiles();
        logger.info(`yt-dlp 路径: ${this.ytdlpPath}`);
        logger.info(`下载目录: ${this.downloadDir}`);
    }

    _findYtdlp() {
        // 优先使用 bin/ 目录下的二进制
        const binDir = path.join(__dirname, '..', 'bin');
        const isWin = process.platform === 'win32';
        const binaryName = isWin ? 'yt-dlp.exe' : 'yt-dlp';

        const localPath = path.join(binDir, binaryName);
        if (fs.existsSync(localPath)) return localPath;

        // 回退到系统 PATH
        return binaryName;
    }

    _cleanupPartialFiles() {
        const patterns = ['.f*.mp4', '.f*.m4a', '.f*.webm', '.ytdl', '.part'];
        try {
            const files = fs.readdirSync(this.downloadDir);
            for (const file of files) {
                if (patterns.some(p => file.includes(p.replace('*', ''))) ||
                    file.endsWith('.ytdl') || file.endsWith('.part')) {
                    try {
                        fs.unlinkSync(path.join(this.downloadDir, file));
                    } catch (e) { /* ignore */ }
                }
            }
        } catch (e) { /* ignore */ }
    }

    getInfo(url, proxy) {
        return new Promise((resolve) => {
            const args = [
                '--dump-json', '--no-download', '--no-warnings',
                '--no-check-certificates',
            ];
            if (proxy) args.push('--proxy', proxy);
            args.push(url);

            const proc = spawn(this.ytdlpPath, args, { timeout: 30000 });
            let stdout = '';
            let stderr = '';

            proc.stdout.on('data', (d) => stdout += d);
            proc.stderr.on('data', (d) => stderr += d);

            proc.on('close', (code) => {
                if (code === 0) {
                    try {
                        const info = JSON.parse(stdout);
                        resolve({
                            title: info.title || '未知',
                            duration: info.duration || 0,
                            thumbnail: info.thumbnail || '',
                            uploader: info.uploader || '未知',
                            view_count: info.view_count || 0,
                        });
                    } catch (e) {
                        resolve({ error: '解析视频信息失败' });
                    }
                } else {
                    resolve({ error: classifyError(stderr || '获取视频信息失败') });
                }
            });

            proc.on('error', (e) => {
                resolve({ error: classifyError(e.message) });
            });
        });
    }

    startDownload(url, proxy, quality = '1080p') {
        const taskId = crypto.randomBytes(4).toString('hex');
        const task = new DownloadTask(taskId, url, quality);
        this.tasks.set(taskId, task);

        this._downloadWorker(task, proxy);
        return taskId;
    }

    getStatus(taskId) {
        const task = this.tasks.get(taskId);
        if (!task) return { error: '任务不存在' };
        return {
            task_id: task.taskId,
            status: task.status,
            progress: Math.round(task.progress * 10) / 10,
            filename: task.filename,
            title: task.title,
            speed: task.speed,
            eta: task.eta,
            error: task.error,
            filesize: task.filesize,
            quality: task.quality,
        };
    }

    listFiles() {
        try {
            const files = fs.readdirSync(this.downloadDir)
                .filter(f => {
                    if (!f.endsWith('.mp4') && !f.endsWith('.webm') && !f.endsWith('.m4a')) return false;
                    if (/\.f\d+\.\w+$/.test(f)) return false;
                    return true;
                })
                .map(f => {
                    const filepath = path.join(this.downloadDir, f);
                    const stats = fs.statSync(filepath);
                    const sizeBytes = stats.size;
                    let size;
                    if (sizeBytes < 1024) size = `${sizeBytes} B`;
                    else if (sizeBytes < 1024 * 1024) size = `${(sizeBytes / 1024).toFixed(1)} KB`;
                    else size = `${(sizeBytes / (1024 * 1024)).toFixed(1)} MB`;
                    return { name: f, size, size_bytes: sizeBytes };
                })
                .sort((a, b) => b.size_bytes - a.size_bytes);
            return files;
        } catch (e) {
            return [];
        }
    }

    activeTaskCount() {
        let count = 0;
        for (const task of this.tasks.values()) {
            if (['pending', 'downloading', 'processing'].includes(task.status)) count++;
        }
        return count;
    }

    _downloadWorker(task, proxy) {
        const fmt = QUALITY_FORMATS[task.quality] || QUALITY_FORMATS['1080p'];
        const cookiesPath = path.join(path.dirname(this.downloadDir), 'cookies.txt');
        const hasCookies = fs.existsSync(cookiesPath);

        const args = [
            '--newline', '--no-warnings',
            '--merge-output-format', 'mp4',
            '-f', fmt,
            '--format-sort', 'vcodec:h264,res,ext:mp4:m4a',
            '-o', path.join(this.downloadDir, '%(title)s.%(ext)s'),
            '--user-agent', 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36',
            '--no-part', '--retries', '30', '--fragment-retries', '30',
            '--retry-sleep', 'linear=1::2', '--no-check-certificates',
            '--socket-timeout', '30', '--extractor-retries', '5',
            '--file-access-retries', '5', '--continue',
        ];

        if (hasCookies) args.push('--cookies', cookiesPath);
        if (proxy) args.push('--proxy', proxy);
        args.push(task.url);

        task.status = 'downloading';
        task.lastProgressTime = Date.now();

        // 超时检测
        task._timeoutTimer = setInterval(() => {
            if (Date.now() - task.lastProgressTime > DOWNLOAD_TIMEOUT) {
                task.status = 'error';
                task.error = '下载超时，可能是网络问题，请检查代理设置后重试';
                clearInterval(task._timeoutTimer);
                proc.kill();
            }
        }, 10000);

        logger.info(`开始下载: ${task.taskId} -> ${task.url} [${task.quality}]`);

        const proc = spawn(this.ytdlpPath, args);
        let buffer = '';

        proc.stdout.on('data', (data) => {
            buffer += data.toString();
            const lines = buffer.split('\n');
            buffer = lines.pop(); // 保留不完整的行
            for (const line of lines) {
                if (line.trim()) this._parseProgress(task, line.trim());
            }
        });

        proc.stderr.on('data', (data) => {
            const line = data.toString().trim();
            if (line) this._parseProgress(task, line);
        });

        proc.on('close', (code) => {
            clearInterval(task._timeoutTimer);
            // 处理剩余 buffer
            if (buffer.trim()) this._parseProgress(task, buffer.trim());

            if (code === 0) {
                task.status = 'completed';
                task.progress = 100;
                this._findDownloadedFile(task);
                logger.info(`下载完成: ${task.taskId} -> ${task.filename}`);
            } else {
                task.status = 'error';
                if (!task.error) {
                    task.error = classifyError(`yt-dlp 返回错误码: ${code}`);
                }
                logger.error(`下载失败: ${task.taskId} -> ${task.error}`);
            }
        });

        proc.on('error', (e) => {
            clearInterval(task._timeoutTimer);
            task.status = 'error';
            task.error = classifyError(e.message);
            logger.error(`下载异常: ${task.taskId} -> ${e.message}`);
        });
    }

    _parseProgress(task, line) {
        // [download] 45.2% of 120.50MiB at 2.30MiB/s ETA 00:30
        let match = line.match(/\[download\]\s+([\d.]+)%\s+of\s+~?\s*(\S+)\s+at\s+(\S+)\s+ETA\s+(\S+)/);
        if (match) {
            task.progress = parseFloat(match[1]);
            task.filesize = match[2];
            task.speed = match[3];
            task.eta = match[4];
            task.status = 'downloading';
            task.lastProgressTime = Date.now();
            return;
        }

        // [download] 100% of 120.50MiB
        match = line.match(/\[download\]\s+([\d.]+)%\s+of\s+~?\s*(\S+)/);
        if (match) {
            task.progress = parseFloat(match[1]);
            task.filesize = match[2];
            task.lastProgressTime = Date.now();
            return;
        }

        // [download] Destination: filename.mp4
        match = line.match(/\[download\] Destination:\s+(.+)/);
        if (match) {
            task.filename = sanitizeFilename(path.basename(match[1]));
            return;
        }

        // [Merger] Merging formats...
        if (line.includes('[Merger]') || line.includes('Merging')) {
            task.status = 'processing';
            task.progress = 99;
            task.lastProgressTime = Date.now();
            return;
        }

        // already been downloaded
        match = line.match(/\[download\]\s+(.+)\s+has already been downloaded/);
        if (match) {
            task.filename = sanitizeFilename(path.basename(match[1]));
            task.status = 'completed';
            task.progress = 100;
            return;
        }

        // ERROR
        if (line.includes('ERROR')) {
            task.error = classifyError(line);
        }
    }

    _findDownloadedFile(task) {
        if (task.filename && !/\.f\d+\.\w+$/.test(task.filename)) return;
        try {
            const files = fs.readdirSync(this.downloadDir)
                .filter(f => f.endsWith('.mp4') && !/\.f\d+\.mp4$/.test(f));
            if (files.length > 0) {
                files.sort((a, b) => {
                    const sa = fs.statSync(path.join(this.downloadDir, a));
                    const sb = fs.statSync(path.join(this.downloadDir, b));
                    return sb.mtimeMs - sa.mtimeMs;
                });
                task.filename = files[0];
            }
        } catch (e) { /* ignore */ }
    }
}

module.exports = { VideoDownloader };
